storage/local/doubledelta.go

```diff
@@ -0,0 +1,513 @@
+		return 1
+	}
+	return (len(c)-doubleDeltaHeaderBytes)/c.sampleSize() + 2
+}
+
+func (c doubleDeltaEncodedChunk) isInt() bool {
+	return c[doubleDeltaHeaderIsIntOffset] == 1
+}
+
+// add implements chunk.
+func (c doubleDeltaEncodedChunk) add(s *metric.SamplePair) []chunk {
+	if len(c) <= doubleDeltaHeaderIsIntOffset+1 {
+		// This is the first sample added to this chunk. Add it as base
+		// time and value.
+		c = c[:doubleDeltaHeaderBaseValueOffset+8]
+		binary.LittleEndian.PutUint64(
+			c[doubleDeltaHeaderBaseTimeOffset:],
+			uint64(s.Timestamp),
+		)
+		binary.LittleEndian.PutUint64(
+			c[doubleDeltaHeaderBaseValueOffset:],
+			math.Float64bits(float64(s.Value)),
+		)
+		return []chunk{&c}
+	}
+
+	tb := c.timeBytes()
+	vb := c.valueBytes()
+
+	if len(c) <= doubleDeltaHeaderBaseValueOffset+8 {
+		// This is the 2nd sample added to this chunk. Calculate the
+		// base deltas.
+		baseTimeDelta := s.Timestamp - c.baseTime()
+		if baseTimeDelta < 0 {
+			// TODO(beorn7): We ignore this irregular case for now. Once
+			// https://github.com/prometheus/prometheus/issues/481 is
+			// fixed, we should panic here instead.
+			return []chunk{&c}
+		}
+		c = c[:doubleDeltaHeaderBytes]
+		if tb >= d8 || bytesNeededForUnsignedTimestampDelta(baseTimeDelta) >= d8 {
+			// If already the base delta needs d8 (or we are at d8
+			// already, anyway), we better encode this timestamp
+			// directly rather than as a delta and switch everything
+			// to d8.
+			c[doubleDeltaHeaderTimeBytesOffset] = byte(d8)
+			binary.LittleEndian.PutUint64(
+				c[doubleDeltaHeaderBaseTimeDeltaOffset:],
+				uint64(s.Timestamp),
+			)
+		} else {
+			binary.LittleEndian.PutUint64(
+				c[doubleDeltaHeaderBaseTimeDeltaOffset:],
+				uint64(baseTimeDelta),
+			)
+		}
+		baseValue := c.baseValue()
+		baseValueDelta := s.Value - baseValue
+		if vb >= d8 || baseValue+baseValueDelta != s.Value {
+			// If we can't reproduce the original sample value (or
+			// if we are at d8 already, anyway), we better encode
+			// this value directly rather than as a delta and switch
+			// everything to d8.
+			c[doubleDeltaHeaderValueBytesOffset] = byte(d8)
+			c[doubleDeltaHeaderIsIntOffset] = 0
+			binary.LittleEndian.PutUint64(
+				c[doubleDeltaHeaderBaseValueDeltaOffset:],
+				math.Float64bits(float64(s.Value)),
+			)
+		} else {
+			binary.LittleEndian.PutUint64(
+				c[doubleDeltaHeaderBaseValueDeltaOffset:],
+				math.Float64bits(float64(baseValueDelta)),
+			)
+		}
+		return []chunk{&c}
+	}
+
+	remainingBytes := cap(c) - len(c)
+	sampleSize := c.sampleSize()
+
+	// Do we generally have space for another sample in this chunk? If not,
+	// overflow into a new one.
+	if remainingBytes < sampleSize {
+		overflowChunks := c.newFollowupChunk().add(s)
+		return []chunk{&c, overflowChunks[0]}
+	}
+
+	dt := s.Timestamp - c.baseTime() - clientmodel.Timestamp(c.len())*c.baseTimeDelta()
+	dv := s.Value - c.baseValue() - clientmodel.SampleValue(c.len())*c.baseValueDelta()
```

Comment:
This is the key point where we need to decide whether we want to guarantee that if you put a valid float64 value into the storage, you get it back precisely as you put it in (in that case we'd have to upgrade to `d8` here if reversing the above calculation doesn't lead us to the original value again), or whether we're ok with small epsilons, as it is currently. Let's discuss tomorrow?